:warning: Add initial Maintainers Annotation parsing

gabibguti opened the pull request:
#### What kind of change does this PR introduce?

Enables maintainers to write annotations for Scorecard checks in a scorecard.yml file.

For example, to provide a reasoning that binaries are present in the code but are only used for tests. Scorecard will read the scorecard.yml file from the repository's root (in the expected format) and parse the annotated checks to display them in scorecards.dev UI and ignore them in GitHub's security alerts.

Design doc: https://docs.google.com/document/d/1-5NKRciF3qU-vLS4xPk48EDfC8isz0Z9vnvL4OVjwpQ/edit#heading=h.xzptrog8pyxf

- [x] PR title follows the guidelines defined in our [pull request documentation](https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process)

#### What is the current behavior?

Scorecard runs checks over a repository.

#### What is the new behavior (if this is a feature change)?**

Maintainers can write annotations to these checks providing a reasoning behind a check's low score.

- [X] Tests for the changes have been added (for bug fixes/features)

#### Which issue(s) this PR fixes

Related to https://github.com/ossf/scorecard/issues/1907

#### Special notes for your reviewer

#### Does this PR introduce a user-facing change?

For user-facing changes, please add a concise, human-readable release note to
the `release-note`

(In particular, describe what changes users might need to make in their
application as a result of this pull request.)

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below.
If the PR requires additional action from users switching to the new release,
include the string "ACTION REQUIRED".

For more information on release notes see: https://git.k8s.io/release/cmd/release-notes/README.md
-->

```release-note
Enables maintainers to write annotations for Scorecard checks and consumers to view these annotations in Scorecard UI.
```

spencerschrock commented on README.md:
````diff
@@ -419,6 +419,60 @@ RESULTS
 |---------|------------------------|--------------------------------|--------------------------------|---------------------------------------------------------------------------|
 ```
 
+##### Showing Maintainers Annotation
+
+To see maintainers annotations for each check, use the `--show-ma` option. The annotations are configured as labels but parsed to a human-readable format in the output.
````
Style question:

We talk about maintainer annotations, but in the code should we just refer to it as `annotations`?

e.g. `package annotations` or cli flag: `--show-annotations`

spencerschrock commented on cmd/root.go:
```diff
@@ -167,6 +167,7 @@ func rootCmd(o *options.Options) error {
 		&repoResult,
 		checkDocs,
 		pol,
+		repoResult.MaintainersAnnotation,
```
we already have `repoResult` as an earlier argument. I don't think we need to change the signature here.

gabibguti replied:
Done.

gabibguti replied:
Thanks for the call out, I forgot to remove it.

spencerschrock commented on README.md:
```diff
@@ -419,6 +419,60 @@ RESULTS
+
+To see maintainers annotations for each check, use the `--show-ma` option. The annotations are configured as labels but parsed to a human-readable format in the output.
+
+To add annotations to your repository, create a `scorecard.yml` file in the root of your repository.
+
```
in an effort to keep the README shorter/higher-level, we may want to move everything here and below to a separate README for detailed maintainer annotation instructions.

gabibguti replied:
Done.

spencerschrock commented on maintainers_annotation/maintainers_annotation.go:
```diff
@@ -0,0 +1,190 @@
+
+// readScorecardYmlFromRepo reads the scorecard.yml file from the repo and returns its configurations.
+func readScorecardYmlFromRepo(repoClient clients.RepoClient) (MaintainersAnnotation, error) {
+	ma := MaintainersAnnotation{}
+	err := fileparser.OnMatchingFileContentDo(repoClient, fileparser.PathMatcher{
+		Pattern:       "scorecard.yml",
+		CaseSensitive: false,
+	}, validateAndParseScorecardYmlFile, &ma)
```
Simpler to use `GetFileContent`? 

This approach would match any file named "scorecard.yml", which could be the Scorecard Action workflow file for example. Which I see you have a check for elsewhere, but if we have specific spots to look, might be easier to just look there.

gabibguti replied:
Done.

gabibguti replied:
Thanks! I was looking for such a function but I could not find it. Works perfectly.

spencerschrock commented on maintainers_annotation/maintainers_annotation.go:
```diff
@@ -0,0 +1,190 @@
+		CaseSensitive: false,
+	}, validateAndParseScorecardYmlFile, &ma)
+
+	if err != nil {
+		return ma, fmt.Errorf("%w", err)
+	}
```
nit: generally you should add context to an error when wrapping, not just %w

gabibguti replied:
Sorry about that, noted.

spencerschrock commented on maintainers_annotation/maintainers_annotation.go:
```diff
@@ -0,0 +1,190 @@
+
+var errNoScorecardYmlFile = errors.New("scorecard.yml doesn't exist or file doesn't contain exemptions")
+
+// AnnotationReason is the reason behind an annotation.
+type AnnotationReason string
+
+const (
+	// AnnotationTestData is to annotate when a check or probe is targeting a danger
+	// in files or code snippets only used for test or example purposes.
+	AnnotationTestData AnnotationReason = "test-data"
```
nit/style:

There's a lot of repetition in the variable names. Callers would access:
```go
maintainers_annotation.AnnotationTestData
```

Recommend 
```go
type Reason string

const (
    TestData Reason = "test-data"
    // etc
```

so it reads better:
```go
maintainers_annotation.TestData
```

gabibguti replied:
Aw, sure. I was sort of copying a pattern from `checks` like `checks.CheckBinaryArtifacts`, but I agree it causes too much repetition, noted.

gabibguti replied:
Done.

spencerschrock commented on maintainers_annotation/maintainers_annotation.go:
```diff
@@ -0,0 +1,190 @@
+	// Read the scorecard.yml file and parse it into maintainers annotation
+	ma, err := readScorecardYmlFromRepo(repoClient)
+
+	// If scorecard.yml doesn't exist or file doesn't contain exemptions, ignore
+	if err == errNoScorecardYmlFile {
```
Good, annotations should be optional. 
We don't want to penalize repos which don't have the file, as the vast majority wont.

gabibguti replied:
Nice.

spencerschrock commented on maintainers_annotation/maintainers_annotation.go:
```diff
@@ -0,0 +1,190 @@
+	return ma, nil
+}
+
+// GetMaintainersAnnotation reads the maintainers annotation from the repo, stored in scorecard.yml, and returns it.
+func GetMaintainersAnnotation(repoClient clients.RepoClient) (MaintainersAnnotation, error) {
```
what do you think about:
```go
func Parse(r io.Reader) (MaintainersAnnotation, error)
```

The annotations don't know need to know about the repo clients, just parse what they're given.

gabibguti replied:
Trying to address this change, I've got:
```diff
# scorecard.go

	// Get configuration
-	c, err := config.Parse(repoClient)
+	content, err := repoClient.GetFileContent("scorecard.yml")
+	// If configuration file exists, continue. Otherwise, ignore
+	if err == nil {
+		c, err := config.Parse(bytes.NewReader(content))
		if err != nil {
			return ScorecardResult{}, err
		}
		ret.Config = c
+	}

```

```diff
# config.go

// Parse reads the configuration file from the repo, stored in scorecard.yml, and returns a `Config`.
- func Parse(repoClient clients.RepoClient) (Config, error) {
+ func Parse(r io.Reader) (Config, error) {
	c := Config{}
-	// Find scorecard.yml file in the repository's root
-	content, err := repoClient.GetFileContent("scorecard.yml")
+	content, err := io.ReadAll(r)
	if err != nil {
-		if errors.Is(err, os.ErrNotExist) {
-			// If scorecard.yml doesn't exist, ignore
-			return c, nil
-		}
		return c, fmt.Errorf("fail to read configuration file: %w", err)
	}

	err = parseFile(&c, content)
	if err != nil {
		return c, fmt.Errorf("fail to parse configuration file: %w", err)
	}

	// Return configuration
	return c, nil
}

```

Is this what you were expecting?

gabibguti replied:
Ack, thanks! And thanks for reminding about closing the file.

gabibguti replied:
Done.

spencerschrock commented on maintainers_annotation/maintainers_annotation.go:
```diff
@@ -0,0 +1,190 @@
+		return MaintainersAnnotation{}, nil
+	}
+	// If an error happened while finding or parsing scorecard.yml file, then raise error
+	if err != nil {
+		return MaintainersAnnotation{}, fmt.Errorf("%w", err)
+	}
```
nit: add context

gabibguti replied:
Done.

spencerschrock commented on maintainers_annotation/maintainers_annotation.go:
```diff
@@ -0,0 +1,190 @@
+	return ma, nil
+}
+
+// IsCheckExempted verifies if a given check in the results is exempted in maintainers annotation.
+func (ma *MaintainersAnnotation) IsCheckExempted(check checker.CheckResult) (bool, []string) {
```
I would be tempted to put this in `checker` package.

spencerschrock commented on maintainers_annotation/maintainers_annotation.go:
```diff
@@ -0,0 +1,190 @@
+	}
+
+	for _, exemption := range ma.Exemptions {
+		for _, checkName := range exemption.Checks {
+			if strings.EqualFold(checkName, strings.ToLower(check.Name)) {
```
`strings.EqualFold` already does case insensitive comparison. 

the call to strings.ToLower isn't needed

gabibguti replied:
Sorry, noted.

gabibguti replied:
Done.

spencerschrock commented on pkg/json.go:
```diff
@@ -49,11 +49,12 @@ type jsonCheckDocumentationV2 struct {
+	Score                 int                      `json:"score"`
+	Reason                string                   `json:"reason"`
+	Name                  string                   `json:"name"`
+	Doc                   jsonCheckDocumentationV2 `json:"documentation"`
+	MaintainersAnnotation []string                 `json:"maintainersAnnotation"`
```
as written, this would break the bigquery table. Need to think about how to not do that. Maybe some form of `omitempty`?

maybe `jsonCheckResultV3`?

spencerschrock replied:
It would add the `maintainersAnnotation` to the JSON output.
https://go.dev/play/p/kwtvfXV61cM

I was thinking if combined with the experimental flag, this would prevent the struct from showing up in any output before we're ready for the feature.

> This will also not impact the cronjob or bigquery table as the JSON results are defined elsewhere.

Agreed. I forgot the cron handles JSON output separately.

gabibguti replied:
> I was thinking if combined with the experimental flag, this would prevent the struct from showing up in any output before we're ready for the feature.

Oh, I think I got it now. Thanks for the playground example! I will check it out and add the `omitempty` field.

gabibguti replied:
Done. Will review when testing.

spencerschrock commented on pkg/json.go:
```diff
@@ -122,7 +123,8 @@ func (r *ScorecardResult) AsJSON(showDetails bool, logLevel log.Level, writer io
 
 // AsJSON2 exports results as JSON for new detail format.
 func (r *ScorecardResult) AsJSON2(showDetails bool,
-	logLevel log.Level, checkDocs docs.Doc, writer io.Writer,
+	showMaintainersAnnotation bool, logLevel log.Level,
+	checkDocs docs.Doc, writer io.Writer,
```
minor:
use a configuration struct maybe? here's a similar one for probe format for example

https://github.com/ossf/scorecard/blob/5a96bddb3aaffeb819d8653051b259dbfd08ffa2/pkg/probe.go#L34-L42

gabibguti replied:
My feeling during the code change is that it didn't get better. It's too verbose and created new types only to create the options structure. I'd rather keep the long function parameters for now and refactor the whole function parameters calls some other time. But if you have a strong preference for this format, I can make the changes.

spencerschrock replied:
part of it is a desire to be able to extend the function in the future in a non-breaking way. Modifying function signatures is a breaking change. So long term, breaking it once means not breaking it again.

it doesn't need to be included in this PR, but is something I see value in

gabibguti replied:
Got it. One question though, what you have in mind is something like:
```
struct AsJSON2ResultOption {
  ShowDetails bool
  ShowMaintainersAnnotation bool
}

func (r *ScorecardResult) AsJSON2(o *AsJSON2ResultOption,
     logLevel log.Level, checkDocs docs.Doc, writer io.Writer)
```
with `AsJSON2ResultOption` encapsulating only the result options or encapsulating other fields too like `logLevel` and `checkDocs`?

spencerschrock replied:
`checkDocs` is fundamental to the call. so I would lean towards something like:
```go
struct AsJSON2Option {
  Details bool
  MaintainersAnnotation bool
  LogLevel log.Level
}

func (r *ScorecardResult) AsJSON2(writer io.Writer, checkDocs docs.Doc, o *AsJSON2Option)
 ```

gabibguti replied:
Ack, thanks for the response!

gabibguti replied:
Done for AsJSON2 and AsString.

spencerschrock commented on maintainers_annotation/maintainers_annotation.go:
```diff
@@ -0,0 +1,190 @@
+	AnnotationNotDetected AnnotationReason = "not-detected"
+)
+
+// AnnotationReasonDoc maps the reasons to their human-readable explanations.
+var AnnotationReasonDoc = map[AnnotationReason]string{
```
public, maps are generally discouraged, as they're world-writable.

raghavkaul replied:
Another way to do this is:
```golang
type AnnotationReason struct {
  Label string
  Doc string
}
```

gabibguti replied:
Changed to a switch approach instead of putting it in a struct. In the struct approach I understand I would still need to instantiate a global obj and switches seemed more performatic. Let me know if you agree with the new approach.

spencerschrock commented on maintainers_annotation/maintainers_annotation.go:
```diff
@@ -0,0 +1,190 @@
+// getAnnotations parses a group of annotations into annotation reasons.
+func getAnnotations(a []Annotation) []string {
+	var annotationReasons []string
+	for _, annotation := range a {
+		annotationReasons = append(annotationReasons, AnnotationReasonDoc[annotation.AnnotationReason])
```
what happens when someone passes an annotation that isn't one of the ones in the map?

gabibguti replied:
Good question. I was planning on solving this while doing the tests. My suggestion is that it should be logged that the reason or check was not identified but not throw an error in the execution, just ignore the weird check / annotation.

gabibguti replied:
Not sure if this makes sense. Another option is to log an error saying it could not parse the annotations and return an empty maintainers annotation to not break the execution.

spencerschrock replied:
generally, logging that we're ignoring/skipping an unknown annotation but reading the rest of the file seems fine

spencerschrock reviewed (commented):
At a high level:
1. the annotation package shouldn't need to know about any of the things above it.
2. I dont want this showing up in the cron yet.

I left a few style comments that can be ignored for now while we discuss approach. I haven't had a chance to play around with the code, just reading it.

pnacht commented on maintainers_annotation/maintainers_annotation.go:
```diff
@@ -0,0 +1,190 @@
+}
+
+// MaintainersAnnotation is a group of exemptions.
+type MaintainersAnnotation struct {
+	Exemptions []Exemption `yaml:"exemptions"`
+}
```
Actually, if the idea is to add more configs (beyond annotating checks) to the .scorecard.yml file in the future*, might it make sense to move all the scorecard.yml reading and parsing to a separate module?

```suggestion
type ScorecardConfig struct {
	MaintainerAnnotations []Exemption `yaml:"exemptions"`
}
```

```go
// config/config.go
import "maintainer_annotations"

func parseScorecardConfigFile(cfg *ScorecardConfig, content[]byte) error {...}
func readScorecardConfigFromRepo(...) (ScorecardConfig, error) {...}
func GetScorecardConfig(...) (ScorecardConfig, error) {...}
```

Whenever the maintainer annotations are needed, they're in scorecardConfig.MaintainerAnnotations.

The maintainer_annotation/ folder would contain all the annotation-specific logic (`IsExempted`, `getAnnotations`).

---

This may be overkill at this very moment (since the ScorecardConfig === MaintainerAnnotations), but if the idea is to add more configs in the future, moving these functions elsewhere later would be a breaking change for any other projects using those functions, like this one:

```go
// github.com/pnacht/scorecard_super_scraper
import ".../scorecard/maintainer_annotation"

func scrapeAllTheReposAndTheirScorecardStuffs() {
    for repo := range GetAllTheRepos() {
        // ... a lot of scraping ...
        GetMaintainerAnnotations()  // breaks if no longer in /maintainer_annotation
        // ...
    }
}
```
---

\* As I'd understood from the design doc and the choice of keeping the annotations within their own `exemptions:` field

gabibguti replied:
@spencerschrock @raghavkaul Any comments on this renaming changes?

pnacht replied:
On my end, LGTM to all that. I do prefer that scorecard.yml format, too (I prefer `annotations:` over `exemptions:`, given the feature is called Maintainer Annotations).

spencerschrock replied:
+1 to Pedro's point of handling the file as a generic settings file so we dont break people later.

Naming is very subjective. I lean towards a simple "config" package.
`config/config.go` would have a `Parse` function, so callers would do `c := config.Parse(foo)`

Depending on how complex the annotation specific logic is, I could see it in `config/annotations.go` or `config/annotations/annotations.go`

Adding `Scorecard` to the various names is redundant in my opinion.

gabibguti replied:
Please review the renamings @pnacht

pnacht commented on maintainers_annotation/maintainers_annotation.go:
```diff
@@ -0,0 +1,190 @@
+	}
+
+	// Exclude scorecard GitHub Action file
+	if strings.Contains(path, ".github") {
```
```suggestion
	if strings.Contains(path, ".github/workflows") {
```

In case we wish to support saving the .scorecard.yml file in .github/.

gabibguti replied:
My understand is that we wanted to support:

File location:
If it's a repository, in the repository root directory
If it's a GitHub organization, in the root directory of the .github repository for the organization

Not sure if this change satisfies the second option or in a normal repository to put the scorecard.yml file in .github/ folder.

gabibguti replied:
Any action items here?

pnacht commented on maintainers_annotation/maintainers_annotation.go:
```diff
@@ -0,0 +1,190 @@
+	}
+
+	if ma.Exemptions == nil {
+		return ma, errNoScorecardYmlFile
```
Is this error (and `if` block) necessary? `if ma.Exemptions == nil`, then `ma == MaintainerAnnotations{}`.

The error is currently used below to just set `ma = MaintainerAnnotations{}` again.

pnacht replied:
Even if the idea is to have more fields in the .scorecard.yml file in the future (i.e. the path to the release workflow), I believe that the lack of a maintainer annotations block in the file will still always lead to `ma = MaintainerAnnotations{}`?

gabibguti replied:
Sorry, I did not understand the issue. The block is necessary to set the proper error and not return this Maintainers Annotations as success case. Once we add more configurations to this file we might want to change this behavior, for example, checking if other "required" fields are missing or not.

pnacht replied:
> The block is necessary to set the proper error and not return this Maintainers Annotations as success case. 

I just don't understand why that is necessary.

The `ma` that is returned here is basically just `MaintainerAnnotations{}`. If you omit this `if` block, you're still going to return the same `MaintainerAnnotations{}` (now without an error).

When this function's output is used by `GetMaintainersAnnotation`, it checks to see if the `errNoScorecardYmlFile` error was raised:
- If it was raised, then it returns `MaintainerAnnotations{}` (which is identical to the `ma` returned by this function.
- If it wasn't raised, then it returns the `ma` returned by this function (which in this case is identical to `MaintainerAnnotations{}`)

So, with this `if` block and associated logic in `GetMaintainerAnnotations`, you ensure that a missing/empty YML file leads to `MaintainerAnnotations{}` being returned in the end.

But if you remove this `if` block (and logic in `GetMaintainerAnnotations`), you already get `MaintainerAnnotations{}`.

gabibguti replied:
Got it, maybe it's indeed unnecessary. Will review after we make a decision on the naming.

raghavkaul commented on checker/check_result.go:
```diff
@@ -262,3 +264,22 @@ func LogFindings(findings []finding.Finding, dl DetailLogger) {
+	// If check has a maximum score, then there it doesn't make sense anymore to reason the check
+	// This may happen if the check score was once low but then the problem was fixed on Scorecard side
+	// or on the maintainers side
+	if check.Score == MaxResultScore {
+		return false, nil
```
If we remove this check it allows repo maintainers to self-report false negatives in Scorecard (i.e. reasons the score shouldn't be 10/10 even if it is). Not required, but just a thought.

gabibguti replied:
Good point. We can do that, but my guess is that the annotation will be currently used to reason low scores, not high scores. So, I would avoid reasoning that for now, see if the feature is being used or not, and then add this later on if the users need it. Do you agree?

raghavkaul replied:
Yeah, sounds good!

raghavkaul reviewed (commented):
Conceptually, it makes more sense to me to have `[]Exemption` with `Check string; Annotation string`, it's a little more verbose but forces the maintainer to be explicit why individual checks are being exempted. This is just a hunch on what will be more usable/readable, so it's not blocking.

gabibguti commented:
> Conceptually, it makes more sense to me to have `[]Exemption` with `Check string; Annotation string`, it's a little more verbose but forces the maintainer to be explicit why individual checks are being exempted. This is just a hunch on what will be more usable/readable, so it's not blocking.

It could be. The reason for having multiple checks bind to multiple annotations is to facilitate 2 scenarios: writing that multiple checks are "justified" by the same reason, and to annotate that multiple reasons apply to a single check too. Since both solutions work, I don't mind changing to this approach if you prefer.

gabibguti commented on maintainers_annotation/maintainers_annotation.go:
```diff
@@ -0,0 +1,190 @@
+}
+
+// parseScorecardYmlFile takes the scorecard.yml file content and returns a `MaintainersAnnotation`.
+func parseScorecardYmlFile(ma *MaintainersAnnotation, content []byte) error {
```
Good suggestion, thanks.

gabibguti replied:
Done.

gabibguti commented on maintainers_annotation/maintainers_annotation.go:
```diff
@@ -0,0 +1,190 @@
+
+// Exemption defines a group of checks that are being annotated for various reasons.
+type Exemption struct {
+	Checks      []string     `yaml:"checks"`
```
Exactly. It also calls the developer attention to the fact that this structure naming is important to marshal/unmarshal.

gabibguti replied:
Any action items here?

pnacht replied:
Not on my end. I just can't resolve the conversation since I'm not a maintainer/triager.

pnacht commented on maintainers_annotation/maintainers_annotation.go:
```diff
@@ -0,0 +1,190 @@
+	AnnotationReason AnnotationReason `yaml:"annotation"`
+}
+
+// Exemption defines a group of checks that are being annotated for various reasons.
+type Exemption struct {
```
(see my comment regarding your suggested change to the scorecard.yml format)

gabibguti replied:
Please review the renamings

gabibguti commented on maintainers_annotation/maintainers_annotation.go:
```diff
@@ -0,0 +1,190 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package maintainers_annotation
```
Please review the renamings

gabibguti commented:
Rebased to get GetFileContent / GetFileReader change.

spencerschrock commented on config/annotations.go:
```diff
@@ -0,0 +1,72 @@
+// Annotation defines a group of checks that are being annotated for various reasons.
+type Annotation struct {
+	Checks  []string      `yaml:"checks"`
+	Reasons []ReasonGroup `yaml:"reasons"`
+}
```
How is a ReasonGroup different than just having?
```go
Reasons []Reason `yaml:"reasons"`
```

spencerschrock replied:
The checks are labeled at the `Annotation` level, so I assumed the probes would be too.

```go
type Annotation struct {
	Checks  []string `yaml:"checks"`
	Reasons []Reason `yaml:"reasons"`
        Probes  []string `yaml:"probes"`
}
```

gabibguti replied:
Aw, yes, that's correct. My mistake.

The proposal initially said to have something like:

```yml
annotations:
  - checks:
      - binary-artifacts
    reasons:
      - reason: test-data
        path:
          - tests/*
      - reason: remediated
        path:
          - bin/source.bin
  - probes:
      - hasDangerousWorkflowScriptInjection
    reasons:
      - reason: run-with-approval
        path:
          - .github/test.yml
```

So `Probes`  would enter in the `Annotation` level, yes. And the `Path` is the field that could enter the `ReasonGroup`. But we can also go for `Reasons []Reason` and put `Path`  in the `Annotation` level too. WDYT?

spencerschrock commented on config/annotations.go:
```diff
@@ -0,0 +1,72 @@
+	Reasons []ReasonGroup `yaml:"reasons"`
+}
+
+// Doc maps a reason to its human-readable explanation.
+func (r *Reason) Doc() string {
```
Having a `String()` method is very common, so we may want to match that name. Though since these descriptions are long, maybe we keep it as `Doc()`?

spencerschrock commented on pkg/sarif.go:
```diff
@@ -626,6 +626,13 @@ func (r *ScorecardResult) AsSARIF(showDetails bool, logLevel log.Level,
 	for _, check := range r.Checks {
 		check := check
+
+		// If check is exempted, skip
+		exempted, _ := check.IsExempted(r.Config)
+		if exempted {
+			continue
+		}
```
nit: the code isn't that complicated/long, so the comment doesnt add much as written.

The comment would be better if you say why we're skipping. In this case something that mentions SARIF is used for code scanning, so we dont want alerts for things the owner knows about?

What do you think?

gabibguti replied:
Makes a lot of sense. Sorry about that. Will fix.

spencerschrock commented on pkg/scorecard_result.go:
```diff
@@ -189,18 +201,19 @@ func (r *ScorecardResult) AsString(showDetails bool, logLevel log.Level,
 		}
-
+		x = append(x, doc)
+		if o.Annotations {
+			_, reasons := row.IsExempted(r.Config)
+			x = append(x, strings.Join(reasons, "\n"))
+		}
```
i'm curious if these newlines break the table?

spencerschrock reviewed (approved):
Overall approach looks good. Obviously tests, fix build, and fix linter

gabibguti commented on README.md:
```diff
@@ -423,6 +423,10 @@ RESULTS
+##### Showing Maintainers Annotations
+
+To see the maintainers annotations for each check, use the `--show-annotations` option.
+
```
Done.

gabibguti commented:
Checkpoint:
- Added validations for check names and reasons when parsing Annotations
- Added tests to:
  - parse Annotations
  - format Scorecard results as String, JSON and SARIF
  - check if Annotations flag to runs only in experimental mode
  - verify if check is exempted

gabibguti commented:
I am not sure if the SARIF output is working as expected. It is clearing the `runs`. Should it clear only the `results`?

spencerschrock reviewed (commented):
didnt dig into tests too much until we answer some questions

spencerschrock commented:
> I am not sure if the SARIF output is working as expected. It is clearing the `runs`. Should it clear only the `results`?

Was `runs` supposed to be `rules`? I don't know enough about SARIF to know whether we should still populate `runs` if there are no relevant `results`.

gabibguti commented on pkg/scorecard.go:
```diff
@@ -164,6 +165,23 @@ func runScorecard(ctx context.Context,
+			checks = append(checks, check)
+		}
+		c, err := config.Parse(rc, checks)
+		if err != nil {
```
Define log level (maybe top level log)

gabibguti commented:
Added the last unresolved comments to issue https://github.com/ossf/scorecard/issues/4048 so we can fix them later after moving on with this first PR of Maintainers Annotation feature.

spencerschrock commented:
/scdiff generate Dependency-Update-Tool,SAST,Maintained,Packaging,Signed-Releases,Branch-Protection,Code-Review,Token-Permissions,CII-Best-Practices,License,Pinned-Dependencies,Dangerous-Workflow,Vulnerabilities,Security-Policy,Fuzzing,CI-Tests,Binary-Artifacts

spencerschrock commented on pkg/scorecard.go:
```diff
@@ -160,6 +162,23 @@ func runScorecard(ctx context.Context,
 	// If the user runs checks
 	go runEnabledChecks(ctx, repo, request, checksToRun, resultsCh)
 
+	// Get configuration
+	rc, err := repoClient.GetFileReader("scorecard.yml")
+	// If configuration file exists, continue. Otherwise, ignore
+	if err == nil {
+		defer rc.Close()
+		checks := []string{}
+		for check := range checksToRun {
+			checks = append(checks, check)
+		}
+		c, err := config.Parse(rc, checks)
+		if err != nil {
+			logger := sclog.NewLogger(sclog.DefaultLevel)
+			logger.Error(err, "parsing configuration file")
+		}
+		ret.Config = c
+	}
+
```
Please wrap this all in an experimental check, otherwise LGTM

spencerschrock replied:
for probes we had something like:
```go
if os.Getenv(options.EnvVarScorecardExperimental) == "1"
```

gabibguti replied:
Ack

gabibguti replied:
Done

oreillymj commented:
For repo's using the github action, can you clarify the config steps.
Do we add 
scorecard.yml to the root of our repo
.scorecard.yml  to the root of our repo (note the dot prefix in the filename)

or add the exemptions configuration into the existing .github/workflows/scorecard.yml that the github action creates.

spencerschrock commented:
> For repo's using the github action, can you clarify the config steps. Do we add scorecard.yml to the root of our repo .scorecard.yml to the root of our repo (note the dot prefix in the filename)
> 
> or add the exemptions configuration into the existing .github/workflows/scorecard.yml that the github action creates.

We are working on clarifying [our documentation](https://github.com/ossf/scorecard/blob/main/config/README.md) before launch at the end of this month.
